From c5b565679d973b85943e88bd6b0e3d9e1e58fd8b Mon Sep 17 00:00:00 2001
From: Valentin Churavy <v.churavy@gmail.com>
Date: Mon, 23 Apr 2018 15:08:13 -0400
Subject: [PATCH] forcefully disable APPLE specific 'feature'

---
 lib/CMakeLists.txt | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/projects/libcxx/lib/CMakeLists.txt b/projects/libcxx/lib/CMakeLists.txt
index aa5ebf156..0d3220dea 100644
--- a/projects/libcxx/lib/CMakeLists.txt
+++ b/projects/libcxx/lib/CMakeLists.txt
@@ -39,10 +39,10 @@ if (LIBCXX_GENERATE_COVERAGE AND NOT LIBCXX_COVERAGE_LIBRARY)
 endif()
 add_library_flags_if(LIBCXX_COVERAGE_LIBRARY "${LIBCXX_COVERAGE_LIBRARY}")
 
-if (APPLE AND (LIBCXX_CXX_ABI_LIBNAME STREQUAL "libcxxabi" OR
-               LIBCXX_CXX_ABI_LIBNAME STREQUAL "default"))
-  set(LIBCXX_OSX_REEXPORT_SYSTEM_ABI_LIBRARY ON)
-endif()
+# if (APPLE AND (LIBCXX_CXX_ABI_LIBNAME STREQUAL "libcxxabi" OR
+#                LIBCXX_CXX_ABI_LIBNAME STREQUAL "default"))
+#   set(LIBCXX_OSX_REEXPORT_SYSTEM_ABI_LIBRARY ON)
+# endif()
 
 if (LIBCXX_ENABLE_STATIC_ABI_LIBRARY)
   add_library_flags("-Wl,--whole-archive" "-Wl,-Bstatic")
-- 
2.17.0

